src/charm.py

```diff
@@ -55,57 +57,65 @@ def __init__(self, *args):
 
         self.framework.observe(self.on.update_status, self._on_update_status)
         self.framework.observe(self.on.get_primary_action, self._on_get_primary_action)
-
-        # handle removal of replica
-        self.framework.observe(self.on.mongodb_storage_detaching, self._on_mongodb_remove_replica)
+        self.framework.observe(
+            self.on.mongodb_storage_detaching, self._on_mongodb_storage_detaching
+        )
         # if a new leader has been elected update hosts of MongoDB
         self.framework.observe(self.on.leader_elected, self._update_hosts)
-        # when a unit departs the replica set update the hosts of MongoDB
-        self.framework.observe(self.on.mongodb_relation_departed, self._update_hosts)
+        self.framework.observe(self.on.mongodb_relation_departed, self._relation_departed)
 
-    def _update_hosts(self, _) -> None:
-        """Update replica set hosts."""
-        # allow leader to reset which hosts are being used
-        if self.unit.is_leader():
-            self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
+    def _update_hosts(self, event) -> None:
+        """Update replica set hosts and remove any unremoved replicas from the config."""
+        if not self.unit.is_leader() and "replset_initialised" not in self._peers.data[self.app]:
+            return
 
-    def _on_mongodb_remove_replica(self, event: ops.charm.StorageDetachingEvent) -> None:
-        """Removes unit from the MongoDB replica set config and steps down primary if necessary.
+        self.process_unremoved_units(event)
+        self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
 
-        Args:
-            event: The triggering storage detaching event.
+    def _on_mongodb_storage_detaching(self, event: ops.charm.StorageDetachingEvent) -> None:
+        """Before storage detaches, allow removing unit to remove itself from the set.
+
+        If the removing unit is primary also allow it to step down and elect another unit as
+        primary while it still has access to its storage.
         """
-        # only remove and reconfigure if all replicas of MongoDB application are in ready state
-        if not self._mongo.all_replicas_ready():
-            self.unit.status = WaitingStatus("waiting to reconfigure replica set")
-            logger.debug("waiting for all replica set hosts to be ready")
+        try:
+            # remove_replset_member retries for one minute in an attempt to resolve race conditions
+            # it is not possible to defer in storage detached.
+            self._mongo.remove_replset_member(self._unit_ip(self.unit))
+        except NotReadyError:
+            logger.info(
+                "Failed to remove %s from replica set, another member is syncing", self.unit.name
+            )
+            event.defer()
+        except PyMongoError as e:
+            logger.info("Deferring reconfigure: error=%r", e)
             event.defer()
```

Comment:
I'm in doubt if these `defer` calls should be here. I'm a little bit confused with this plus the above comment and the PR description, both stating that isn't allowed to defer on storage detached event.

Reply:
You're right, this is outdated, I did something wrong with git here. Updating this right away. thank you for catching Marcelo :)